### PR TITLE
Improvements to Container CD

### PR DESCRIPTION
**Pull request by tylertitsworth**
## Description

Update CD to follow DevOps best practices.

### How it works

Given an example(s) (AudioQnA), this workflow creates jobs that perform container CI on that example, and allow other event-based workflows to curate a matrix of these examples.

*These are the inputs for the composites:*

#### Build

```yaml
# .github/workflows/composite/docker-build/action.yml
inputs:
  example_dir:
    description: Directory with docker-compose.yaml to build
    required: true
    type: string
  env_overrides:
    description: Bash Env Variable Overrides in `KEY=VAL && KEY2=VAL2` format
    required: false
    type: string
  registry:
    description: Container Registry URL
    required: false
    default: 'opea-project'
    type: string
```

#### Scan

```yaml
# .github/workflows/composite/scan/action.yml
inputs:
  image-ref:
    description: 'image reference(for backward compatibility)'
    required: true
  output:
    description: 'writes results to a file with the specified file name'
    required: true
```

*This is the inputs for the reusable workflow:*

#### Container CI

```yaml
# .github/workflows/container.ci.yaml
inputs:
  example_dir:
    required: true
    type: string
  scan:
    default: true
    required: false
    type: boolean
  test:
    default: true
    required: false
    type: boolean
  publish:
    default: false
    required: false
    type: boolean
```

*These are the Events that this PR supports:*

```yaml
# .github/workflows/example-weekly-ci.yaml
on:
  # weekly
  schedule:
  - cron: "0 0 * * 0"
  # run all examples manually
  workflow_dispatch: null
  push:
    tags: ['**'] # whenever a tag is created/modified
```

```yaml
# ./github/workflows/example-manual-ci.yaml
on:
  workflow_dispatch:
    inputs:
      examples:
        default: 'example1, example2'
        description: 'List of comma-separated examples to test'
        required: true
        type: string
      scan:
        default: true
        description: 'Scan all examples with Trivy'
        required: false
        type: boolean
      test:
        default: true
        description: 'Test Examples'
        required: false
        type: boolean
      publish:
        default: false
        description: 'Publish Images to Dockerhub'
        required: false
        type: boolean
```

#### Changes

- Add StepSecurity network traffic monitoring during workflows via `harden-runner`
- Tests labels are determined based on the test workflow filename:

```bash
# .github/workflows/container-ci.yaml
- name: Get Tests
  id: test-matrix
  run: |
    echo "matrix=$(find ${{ inputs.example_dir }}/tests -type f -name 'test_*.sh' -print | \
    sed -n 's|.*/\(test_[^_]*_on_\([^\.]*\)\).sh|\{"test": "\1", "hardware_label": "\2"}|p' | \
    jq -sc .)" >> $GITHUB_OUTPUT
  # ex: [{"test":"test_audioqna_on_xeon","hardware_label":"xeon"},{"test":"test_audioqna_on_gaudi","hardware_label":"gaudi"}]
...
runs-on: ${{ matrix.tests.hardware_label }}
...
- name: Run Test
  run: bash ${{ matrix.tests.test }}.sh
  env:
    REGISTRY: ${{ secrets.REGISTRY }}
```

- Simplify developer workflow:

```bash
# Before
docker build -t container1
docker build -t container2
docker build -t container3
...
# After
docker compose build
```

- Test only the latest version of that example's container
  - We pull the last known good version of the comps, and test the changes made in the example only.
  - Modifying the test file or docker-compose.yaml file for a given test can allow for testing specific versions of comp containers. No need to modify CI.
- `status-check` flag in `pull_request` events that verifies if any job in `example-ci.yaml` failed.
#### 

## Issues

n/a

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

n/a

## Tests

I don't have the runners required to do the validation portion, however when previously testing with `ubuntu-latest` I was able to launch the test script from the correct directory on that runner before it ran out of space.
https://github.com/tylertitsworth/GenAIExamples/actions/runs/9896085740?pr=3

### CODE REVIEW

File: Translation/docker/gaudi/README.md [hunk 1]
```diff
@@ -17,23 +17,14 @@ cd GenAIComps
...
-Construct the frontend Docker image using the command below:
-
-```bash
-cd GenAIExamples/Translation/docker/ui/
-docker build -t opea/translation-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/Translation
+docker compose build translation translation-ui
 ```
 
 Then run the command `docker images`, you will have the following four Docker Images:
```
Comment (claynerobison) [inline Translation/docker/gaudi/README.md:30]:
```suggestion
Then run the command `docker images`, you will have the following six Docker Images:
```

**Review by claynerobison — CHANGES_REQUESTED** [review]
All I see is the minor doc change. other than that, seems like an elegant architecture improvement.

**Comment by chensuyue** [comment]
How does this new structure compiled with current compose and k8s functionality test? Our test machine are not able to access an internal docker registry, so we have 2 local registry for CI test. We need to sync about the design.

File: AudioQnA/docker/xeon/README.md [hunk 2]
```diff
@@ -34,14 +34,14 @@ docker build -t opea/speecht5:latest --build-arg https_proxy=$https_proxy --buil
...
 
-To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `audioqna.py` Python script. Build the MegaService Docker image using the command below:
+To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `audioqna.py` Python script. Build the MegaService Docker images using the command below:
 
 ```bash
 git clone https://github.com/opea-project/GenAIExamples.git
-cd GenAIExamples/AudioQnA/docker
-docker build --no-cache -t opea/audioqna:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
+cd GenAIExamples/AudioQnA
+docker compose build audioqna audioqna-ui
```
Comment (ashahba) [inline AudioQnA/docker/xeon/README.md:44]: If the goal is to build all micro-services that have a build target, why not just?
```
docker compose build
```
and leave the service names out.

File: ChatQnA/docker/aipc/README.md [hunk 3]
```diff
@@ -44,25 +44,14 @@ docker build --no-cache -t opea/dataprep-redis:latest -f comps/dataprep/redis/la
...
-### 7. Build UI Docker Image
-
-Build frontend Docker image via below command:
-
-```bash
-cd GenAIExamples/ChatQnA/docker/ui/
-docker build --no-cache -t opea/chatqna-ui:latest -f ./docker/Dockerfile .
-cd ../../../..
+cd GenAIExamples/ChatQnA
+docker compose build chatqna chatqna-ui
```
Comment (ashahba) [inline ChatQnA/docker/aipc/README.md:54]: same here with:
```
docker compose build
```

File: ChatQnA/docker/gaudi/README.md [hunk 4]
```diff
@@ -54,25 +54,14 @@ docker build --no-cache -f Dockerfile-hpu -t opea/tei-gaudi:latest .
...
-### 9. Build UI Docker Image
-
-Construct the frontend Docker image using the command below:
-
-```bash
-cd GenAIExamples/ChatQnA/docker/ui/
-docker build --no-cache -t opea/chatqna-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
-cd ../../../..
+cd GenAIExamples/ChatQnA
+docker compose build chatqna chatqna-ui
```
Comment (ashahba) [inline ChatQnA/docker/gaudi/README.md:64]: and here

File: ChatQnA/docker/gpu/README.md [hunk 5]
```diff
@@ -43,25 +43,14 @@ docker build --no-cache -t opea/llm-tgi:latest --build-arg https_proxy=$https_pr
...
-### 8. Build UI Docker Image
-
-Construct the frontend Docker image using the command below:
-
-```bash
-cd GenAIExamples/ChatQnA/docker/ui/
-docker build --no-cache -t opea/chatqna-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
-cd ../../../..
+cd GenAIExamples/ChatQnA
+docker compose build chatqna chatqna-ui
```
Comment (ashahba) [inline ChatQnA/docker/gpu/README.md:53]: and here

File: ChatQnA/docker/xeon/README.md [hunk 6]
```diff
@@ -97,40 +97,20 @@ docker build --no-cache -t opea/dataprep-redis:latest --build-arg https_proxy=$h
...
-To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `chatqna.py` Python script. Build MegaService Docker image via below command:
+To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `chatqna.py` Python script. Build MegaService Docker images via below command:
 
 ```bash
 git clone https://github.com/opea-project/GenAIExamples.git
-cd GenAIExamples/ChatQnA/docker
-docker build --no-cache -t opea/chatqna:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
-cd ../../..
+cd GenAIExamples/ChatQnA
+docker compose build chatqna chatqna-ui
```
Comment (ashahba) [inline ChatQnA/docker/xeon/README.md:107]: and here

File: ChatQnA/docker/xeon/README.md [hunk 7]
```diff
@@ -220,7 +200,7 @@ curl ${host_ip}:6006/embed \
 2. Embedding Microservice
 
 ```bash
-curl http://${host_ip}:6000/v1/embeddings\
+curl http://${HOST_IP}:6000/v1/embeddings\
```
Comment (ashahba) [inline ChatQnA/docker/xeon/README.md:203]: add one space before `\` at the end

File: ChatQnA/docker/xeon/README.md [hunk 8]
```diff
@@ -256,7 +236,7 @@ curl http://${host_ip}:8808/rerank \
 5. Reranking Microservice
 
 ```bash
-curl http://${host_ip}:8000/v1/reranking\
+curl http://${HOST_IP}:8000/v1/reranking\
```
Comment (ashahba) [inline ChatQnA/docker/xeon/README.md:239]: add one space before `\` at the end

File: ChatQnA/docker/xeon/README.md [hunk 9]
```diff
@@ -274,7 +254,7 @@ curl http://${host_ip}:9009/generate \
 7. LLM Microservice
 
 ```bash
-curl http://${host_ip}:9000/v1/chat/completions\
+curl http://${HOST_IP}:9000/v1/chat/completions\
```
Comment (ashahba) [inline ChatQnA/docker/xeon/README.md:257]: add one space before `\` at the end

File: CodeGen/docker/gaudi/README.md [hunk 10]
```diff
@@ -19,23 +19,14 @@ cd GenAIComps
...
-
-### 4. Build the UI Docker Image
-
-Construct the frontend Docker image via the command below:
-
-```bash
-cd GenAIExamples/CodeGen/docker/ui/
-docker build -t opea/codegen-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/CodeGen
+docker compose build codegen codegen-ui
```
Comment (ashahba) [inline CodeGen/docker/gaudi/README.md:29]: same here with:
```
docker compose build
```

File: CodeGen/docker/xeon/README.md [hunk 11]
```diff
@@ -27,23 +27,14 @@ cd GenAIComps
...
-
-### 4. Build the UI Docker Image
-
-Build the frontend Docker image via the command below:
-
-```bash
-cd GenAIExamples/CodeGen/docker/ui/
-docker build -t opea/codegen-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/CodeGen
+docker compose build codegen codegen-ui
```
Comment (ashahba) [inline CodeGen/docker/xeon/README.md:37]: and here with:
```
docker compose build
```

File: CodeTrans/docker/gaudi/README.md [hunk 12]
```diff
@@ -19,19 +19,12 @@ cd GenAIComps
...
-docker build -t opea/codetrans:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
-```
-
-### 4. Build UI Docker Image
-
-```bash
-cd GenAIExamples/CodeTrans/docker/ui
-docker build -t opea/codetrans-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/CodeTrans
+docker compose build codetrans codetrans-ui
```
Comment (ashahba) [inline CodeTrans/docker/gaudi/README.md:27]:
```
docker compose build
```

File: CodeTrans/docker/xeon/README.md [hunk 13]
```diff
@@ -27,19 +27,12 @@ cd GenAIComps
...
-docker build -t opea/codetrans:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
-```
-
-### 4. Build UI Docker Image
-
-```bash
-cd GenAIExamples/CodeTrans/docker/ui
-docker build -t opea/codetrans-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/CodeTrans
+docker compose build codetrans codetrans-ui
```
Comment (ashahba) [inline CodeTrans/docker/xeon/README.md:35]:
```
docker compose build
```

File: DocSum/docker/gaudi/README.md [hunk 14]
```diff
@@ -25,23 +25,14 @@ docker pull ghcr.io/huggingface/tgi-gaudi:1.2.1
...
-
-### 4. Build UI Docker Image
-
-Construct the frontend Docker image using the command below:
-
-```bash
-cd GenAIExamples/DocSum/docker/ui/
-docker build -t opea/docsum-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/DocSum
+docker compose build docsum docsum-ui
```
Comment (ashahba) [inline DocSum/docker/gaudi/README.md:35]:
```
docker compose build
```

File: DocSum/docker/xeon/README.md [hunk 15]
```diff
@@ -27,23 +27,14 @@ docker build -t opea/llm-docsum-tgi:latest --build-arg https_proxy=$https_proxy
...
-
-### 3. Build UI Docker Image
-
-Build the frontend Docker image via below command:
-
-```bash
-cd GenAIExamples/DocSum/docker/ui/
-docker build -t opea/docsum-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/DocSum
+docker compose build docsum docsum-ui
```
Comment (ashahba) [inline DocSum/docker/xeon/README.md:37]:
```
docker compose build
```

File: SearchQnA/docker/gaudi/README.md [hunk 16]
```diff
@@ -48,22 +48,14 @@ docker build --no-cache -f Dockerfile-hpu -t opea/tei-gaudi:latest .
...
-cd ../../..
-```
-
-Then you need to build the last Docker image `opea/searchqna:latest`, which represents the Mega service through following commands:
-
-```bash
-cd GenAIExamples/SearchQnA/docker
-docker build --no-cache -t opea/searchqna:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
+cd GenAIExamples/SearchQnA
+docker compose build searchqna searchqna-ui
```
Comment (ashahba) [inline SearchQnA/docker/gaudi/README.md:58]:
```
docker compose build
```

File: SearchQnA/docker/xeon/README.md [hunk 17]
```diff
@@ -35,15 +35,14 @@ docker build --no-cache -t opea/reranking-tei:latest --build-arg https_proxy=$ht
...
-To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `searchqna.py` Python script. Build the MegaService Docker image using the command below:
+To construct the Mega Service, we utilize the [GenAIComps](https://github.com/opea-project/GenAIComps.git) microservice pipeline within the `searchqna.py` Python script. Build the MegaService Docker images using the command below:
 
 ```bash
 git clone https://github.com/opea-project/GenAIExamples.git
-cd GenAIExamples/SearchQnA/docker
-docker build --no-cache -t opea/searchqna:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
-cd ../../..
+cd GenAIExamples/SearchQnA
+docker compose build searchqna searchqna-ui
```
Comment (ashahba) [inline SearchQnA/docker/xeon/README.md:45]:
```
docker compose build
```

File: SearchQnA/docker-compose.yaml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+services:
+  translation:
+    build:
+      args:
+        http_proxy: ${http_proxy}
+        https_proxy: ${https_proxy}
+        no_proxy: ""
+      context: ./docker
+      dockerfile: Dockerfile
+    image: ${REGISTRY:-opea}/translation:${GITHUB_RUN_NUMBER:-latest}
```
Comment (ashahba) [inline SearchQnA/docker-compose.yaml:14]: You mean:
```
image: ${REGISTRY:-opea}/searchqna:${GITHUB_RUN_NUMBER:-latest}
```

File: SearchQnA/docker-compose.yaml [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+  translation-ui: # This image isn't being utilized, but does exist.
+    build:
+      context: ./docker/ui
+      dockerfile: ./docker/Dockerfile
+    extends: translation
+    image: ${REGISTRY:-opea}/translation-ui:${GITHUB_RUN_NUMBER:-latest}
```
Comment (ashahba) [inline SearchQnA/docker-compose.yaml:20]: You mean:
```
image: ${REGISTRY:-opea}/searchqna-ui:${GITHUB_RUN_NUMBER:-latest}
```

File: Translation/docker/gaudi/README.md [hunk 20]
```diff
@@ -17,23 +17,14 @@ cd GenAIComps
...
-
-### 3. Build UI Docker Image
-
-Construct the frontend Docker image using the command below:
-
-```bash
-cd GenAIExamples/Translation/docker/ui/
-docker build -t opea/translation-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/Translation
+docker compose build translation translation-ui
```
Comment (ashahba) [inline Translation/docker/gaudi/README.md:27]:
```
docker compose build
```

File: Translation/docker/xeon/README.md [hunk 21]
```diff
@@ -25,30 +25,23 @@ cd GenAIComps
...
-
-### 3. Build UI Docker Image
-
-Build frontend Docker image via below command:
-
-```bash
-cd GenAIExamples/Translation/docker/ui
-docker build -t opea/translation-ui:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f ./docker/Dockerfile .
+cd GenAIExamples/Translation
+docker compose build translation translation-ui
```
Comment (ashahba) [inline Translation/docker/xeon/README.md:35]:
```
docker compose build
```

**Review by ashahba — CHANGES_REQUESTED** [review]
Thanks @tylertitsworth for the PR.
My change requests are mostly minor except for introducing dependency on `ai-containers/.github@main` which needs to be reviewed by the community.

**Comment by tylertitsworth** [comment]
> How does this new structure combined with current compose and k8s functionality test? Our test machine are not able to access an internal docker registry, so we have 2 local registry for CI test. We need to sync about the design. 

@chensuyue the current compose structure could be merged with this design. It's all about having a consistent file system and making sure your dependencies are in subfolders so you can accurately track changes to individual containers.

Regarding the local CI, we can change the runner labels and just make sure to use the correct secrets.

Regarding k8s, I believe we previously spoke about how to best test helm charts. We have a composite action for this. However I see a lot of manifests so it might be better to utilize the baremetal test runner mode to call your existing test scripts to validate those manifests. Will try this out tomorrow.

**Comment by tylertitsworth** [comment]
> My change requests are mostly minor except for introducing dependency on `ai-containers/.github@main` which needs to be reviewed by the community.

Thanks @ashahba, it's common practice to re-use GitHub actions from other locations. I hope that the community will accept another intel repository as a trusted and maintained source similar to how they trust GitHub, docker, helm, pre-commit.ci, and other open source tools already in use.

If the fear of a dependency failing is important. These actions can be pinned at a specific SHA to avoid failures like I did with many other actions introduced by this workflow.

File: .github/workflows/reuse-container-ci.yaml [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+      fail-fast: false
+    steps:
+    - name: Harden Runner
+      uses: step-security/harden-runner@17d0e2bd7d51742c71671bd19fa12bdc9d40a3d6 # v2.8.1
+      with:
+        egress-policy: audit
+    - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
+    - uses: docker/login-action@0d4c9c5ea7693da7b068278f7b52bda2a190a446 # v3.2.0
+      with:
+        registry: ${{ secrets.REGISTRY }}
```
Comment (chensuyue) [inline .github/workflows/reuse-container-ci.yaml:89]: How to use a common registry on IDC machine? Now we have 2 local registry separately for xeon and gaudi cluster.

Comment (tylertitsworth) [reply .github/workflows/reuse-container-ci.yaml:89]: If you can use something like harbor that can be accessed by both clusters or a cloud registry that would be ideal.

If your runners have pre-authenticated access to these local registries, then we could remove the login commands/secrets and use environment variables that are present in the runner environment.

**Pull request closed without merging**